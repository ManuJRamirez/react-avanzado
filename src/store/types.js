export const AUTH_LOGIN = 'auth/login';
export const AUTH_LOGOUT = 'auth/logout';

export const GET_TAGS = 'GET_TAGS';
export const GET_AD_LIST = 'GET_AD_LIST';
export const GET_AD_DETAILS = 'GET_AD_DETAILS';
export const CREATE_AD = 'CREATE_AD';
export const DELETE_AD = 'DELETE_AD';

export const FETCH_DATA_REQUEST = 'FETCH_DATA_REQUEST';
export const FETCH_DATA_SUCCESS = 'FETCH_DATA_SUCCESS';
export const FETCH_DATA_FAILURE = 'FETCH_DATA_FAILURE';
