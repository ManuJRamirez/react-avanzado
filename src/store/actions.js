import {
  AUTH_LOGIN,
  AUTH_LOGOUT,
  GET_TAGS,
  GET_AD_LIST,
  GET_AD_DETAILS,
  CREATE_AD,
  DELETE_AD,
  FETCH_DATA_REQUEST,
  FETCH_DATA_SUCCESS,
  FETCH_DATA_FAILURE,
} from './types';

const fetchDataRequest = () => ({ type: FETCH_DATA_REQUEST });
const fetchDataSuccess = data => ({ type: FETCH_DATA_SUCCESS, payload: data });
const fetchDataFailure = error => ({
  type: FETCH_DATA_FAILURE,
  payload: error,
});

export const authLogin = user => ({
  type: AUTH_LOGIN,
  payload: user,
});

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const getTagsData = tags => ({
  type: GET_TAGS,
  payload: { tags },
});

export const getAdList = ads => ({
  type: GET_AD_LIST,
  payload: { ads },
});

export const getAdDetails = detail => ({
  type: GET_AD_DETAILS,
  payload: { detail },
});

export const createAd = adData => ({
  type: CREATE_AD,
  payload: adData,
});

export const deleteAd = adId => ({
  type: DELETE_AD,
  payload: adId,
});

export const fetchDataAsync = url => {
  return async dispatch => {
    dispatch(fetchDataRequest());

    try {
      const response = await fetch(url);
      const data = await response.json();
      dispatch(fetchDataSuccess(data));
    } catch (error) {
      dispatch(fetchDataFailure(error));
    }
  };
};
