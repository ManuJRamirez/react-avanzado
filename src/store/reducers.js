import {
  AUTH_LOGIN,
  AUTH_LOGOUT,
  GET_TAGS,
  GET_AD_LIST,
  GET_AD_DETAILS,
  CREATE_AD,
  DELETE_AD,
} from './types';

const defaultState = {
  auth: false,
  user: null,
  ads: [],
  tags: [],
  detail: {},
};

export function reducer(state = defaultState, action) {
  switch (action.type) {
    case AUTH_LOGIN:
      return {
        ...state,
        user: action.payload,
        auth: true,
      };
    case AUTH_LOGOUT:
      return {
        ...state,
        user: null,
        auth: false,
      };
    case GET_TAGS:
      const { tags } = action.payload;
      return {
        ...state,
        tags: tags,
      };
    case GET_AD_LIST:
      const { ads } = action.payload;
      return {
        ...state,
        ads: ads,
      };
    case GET_AD_DETAILS:
      const { detail } = action.payload;
      return {
        ...state,
        detail: detail,
      };
    case CREATE_AD:
      return {
        ...state,
        ads: [...state.ads, action.payload],
      };
    case DELETE_AD:
      return {
        ...state,
        ads: state.ads.filter(ad => ad.id !== action.payload),
      };
    default:
      return state;
  }
}
