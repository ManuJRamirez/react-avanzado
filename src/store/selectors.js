export const getAuthState = state => state.auth;

export const getUser = state => getAuthState(state).user;

export const getIsLogged = state => getAuthState(state).auth;

export const getAds = state => state.ads;

export const getAdDetailsById = (state, adId) =>
  state.ads.find(ad => ad.id === adId);

export const getTags = state => state.tags;
