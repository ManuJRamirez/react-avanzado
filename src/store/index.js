import { createStore } from 'redux';
import { reducer } from './reducers';

const getInitialState = () => {
  const localStorageState = localStorage.getItem('reduxState');
  return localStorageState ? JSON.parse(localStorageState) : undefined;
};

const store = createStore(
  reducer,
  getInitialState(),
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);

store.subscribe(() => {
  const reduxState = store.getState();
  localStorage.setItem('reduxState', JSON.stringify(reduxState));
});

export default store;
