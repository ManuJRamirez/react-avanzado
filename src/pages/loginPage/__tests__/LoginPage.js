import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import LoginPage from '../loginPage';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { authLogin } from '../../../store/actions';

jest.mock('../../../store/actions');

const userType = (input, text) => userEvent.type(input, text);

describe('LoginPage', () => {
  const state = { ui: { isFetching: false, error: null } };
  const store = {
    getState: () => state,
    subscribe: () => {},
  };

  const renderComponent = () =>
    render(
      <Router>
        <Provider store={store}>
          <LoginPage />
        </Provider>
      </Router>,
    );

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action', async () => {
    const email = 'hola@hola.es';
    const password = 'Hola.123';

    renderComponent();

    const emailInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByTestId('loginButton');

    await act(() => userType(emailInput, email));
    await act(() => userType(passwordInput, password));

    expect(submitButton).toBeEnabled();

    await userEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email });
  });
});
