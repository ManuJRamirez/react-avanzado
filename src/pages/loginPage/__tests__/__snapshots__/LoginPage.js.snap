// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in NodePoP
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <label>
        Recordar contrase√±a:
        <input
          name="rememberMe"
          type="checkbox"
        />
      </label>
      <button
        class="sc-bdfCDU bWAtCh loginForm-submit"
        data-testid="loginButton"
        disabled=""
        type="submit"
      >
        Log in
      </button>
      <button
        class="sc-bdfCDU bWAtCh loginForm-submit"
        data-testid="signUpButton"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</div>
`;
